Unmarshalling error

**Describe the bug**
A clear and concise description of what the bug is.

**To Reproduce**
Steps to reproduce the behavior:
1. ```go run github.com/onosproject/onos-config/cmd/onos-config     -configStore=$HOME/go/src/github.com/onosproject/onos-config/configs/configStore-sample.json     -changeStore=$HOME/go/src/github.com/onosproject/onos-config/configs/changeStore-sample.json     -deviceStore=$HOME/go/src/github.com/onosproject/onos-config/configs/deviceStore-sample.json     -networkStore=$HOME/go/src/github.com/onosproject/onos-config/configs/networkStore-sample.json     -modelPlugin=$HOME/go/src/github.com/onosproject/onos-config/modelplugin/TestDevice-1.0.0/testdevice.so.1.0.0     -modelPlugin=$HOME/go/src/github.com/onosproject/onos-config/modelplugin/TestDevice-2.0.0/testdevice.so.2.0.0     -modelPlugin=$HOME/go/src/github.com/onosproject/onos-config/modelplugin/Devicesim-1.0.0/devicesim.so.1.0.0```
2. ```gnmi_cli -address localhost:5150 -set     -proto "update: <path: <target: 'localhost-1', elem: <name: 'openconfig-system:system'> elem: <name: 'clock' > elem: <name: 'config'> elem: <name: 'timezone-name'>> val: <string_val: 'Europe/Paris'>>"     -timeout 5s -alsologtostderr     -client_crt pkg/southbound/testdata/client1.crt     -client_key pkg/southbound/testdata/client1.key     -ca_crt pkg/southbound/testdata/onfca.crt```
3. See error

**Expected behavior**
A clear and concise description of what you expected to happen.

**Screenshots**
2019/06/14 16:11:48 Added change 8n0PlTklkfRHM3vDlWpHqvcnQSA= to ChangeStore (in memory)
panic: Cannot unmarshal JSON: unmarshalLeafList for schema authentication-method: value openconfig-aaa-types:LOCAL (string): got type string, expect []interface{}

goroutine 26 [running]:
github.com/onosproject/onos-config/modelplugin/Devicesim-1%2e0%2e0.modelplugin.UnmarshalConfigValues(0x0, 0x0, 0xc000101000, 0x39a, 0x400, 0xc000210540, 0x45479a0, 0xc000210540)
	/Users/andrea/go/src/github.com/onosproject/onos-config/modelplugin/Devicesim-1.0.0/modelmain.go:45 +0x18e
github.com/onosproject/onos-config/pkg/manager.(*Manager).SetNetworkConfig(0x4aea4e0, 0xc0000e8180, 0x11, 0xc0002103c0, 0x0, 0x0, 0x0, 0x0, 0xc0000f33a0, 0x442c0eb, ...)
	/Users/andrea/go/src/github.com/onosproject/onos-config/pkg/manager/setconfig.go:117 +0xb27
github.com/onosproject/onos-config/pkg/northbound/gnmi.(*Server).buildUpdateResults(0xc000176e80, 0xc0000f3828, 0xc0002d77f8, 0x0, 0x0, 0x0, 0x0, 0xc0002e4038, 0x0, 0x0, ...)
	/Users/andrea/go/src/github.com/onosproject/onos-config/pkg/northbound/gnmi/set.go:170 +0x1e9
github.com/onosproject/onos-config/pkg/northbound/gnmi.(*Server).Set(0xc000176e80, 0x46ce900, 0xc0002100f0, 0xc000452000, 0x0, 0x0, 0x0)
	/Users/andrea/go/src/github.com/onosproject/onos-config/pkg/northbound/gnmi/set.go:85 +0xad0
github.com/onosproject/onos-config/vendor/github.com/openconfig/gnmi/proto/gnmi._GNMI_Set_Handler(0x45a6960, 0xc000176e80, 0x46ce900, 0xc0002100f0, 0xc000562370, 0x0, 0x46ce900, 0xc0002100f0, 0xc00001e000, 0x67)
	/Users/andrea/go/src/github.com/onosproject/onos-config/vendor/github.com/openconfig/gnmi/proto/gnmi/gnmi.pb.go:2671 +0x23e
github.com/onosproject/onos-config/vendor/google.golang.org/grpc.(*Server).processUnaryRPC(0xc00020e780, 0x46d4620, 0xc00042d800, 0xc000278300, 0xc000210a80, 0x4ade450, 0x0, 0x0, 0x0)
	/Users/andrea/go/src/github.com/onosproject/onos-config/vendor/google.golang.org/grpc/server.go:972 +0x470
github.com/onosproject/onos-config/vendor/google.golang.org/grpc.(*Server).handleStream(0xc00020e780, 0x46d4620, 0xc00042d800, 0xc000278300, 0x0)
	/Users/andrea/go/src/github.com/onosproject/onos-config/vendor/google.golang.org/grpc/server.go:1252 +0xda6
github.com/onosproject/onos-config/vendor/google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc0000c76e0, 0xc00020e780, 0x46d4620, 0xc00042d800, 0xc000278300)
	/Users/andrea/go/src/github.com/onosproject/onos-config/vendor/google.golang.org/grpc/server.go:691 +0x9f
created by github.com/onosproject/onos-config/vendor/google.golang.org/grpc.(*Server).serveStreams.func1
	/Users/andrea/go/src/github.com/onosproject/onos-config/vendor/google.golang.org/grpc/server.go:689 +0xa1
exit status 2

This happens because the validation of the Devicesim configurations are not done until the first attempt is made to change an attribute on one of the devices like this - and then it all blows up :-(

Even when validation is done at startup there are still 2 problems that need to be tackled:
1) The json generated represents the **authentication-method** attribute like 
```
"authentication-method": "openconfig-aaa-types:LOCAL"
```
but because it is a **leaf list** the value should be like
```
"authentication-method": ["openconfig-aaa-types:LOCAL"]
```

For this reason the YGot Marshaller throws this error: 
```
panic: Cannot unmarshal JSON: unmarshalLeafList for schema authentication-method: value openconfig-aaa-types:LOCAL (string): got type string, expect []interface{}
```

The question is how to deal with this type of thing - there is no way of knowing from a simple string value that the type should be leaf-list - this can only be inferred from the Schema.

Further investigation is needed. This is not the only such case - the **aux-id** expects to be a **float64**

--- a/cmd/onos-config/onos-config.go
+++ b/cmd/onos-config/onos-config.go
@@ -112,6 +112,11 @@ func main() {
 				log.Fatal("Unable to start onos-config ", err)
 			}
 		}
+		err = mgr.ValidateStores()
+		if err != nil {
+			log.Error("Error when validating configurations against model plugins: ", err)
+			os.Exit(-1)
+		}
 
 		mgr.Run()
 		err = startServer(*caPath, *keyPath, *certPath)

--- a/configs/changeStore-sample.json
+++ b/configs/changeStore-sample.json
@@ -197,14 +197,14 @@
         }
       ]
     },
-    "i86kLjjNFr5IBa+O3YTuv9HBqJg=": {
-      "ID": "i86kLjjNFr5IBa+O3YTuv9HBqJg=",
+    "UwIDulssBgfXoeKco0eirkKkJUI=": {
+      "ID": "UwIDulssBgfXoeKco0eirkKkJUI=",
       "Description": "Modification of device sim 1",
       "Created": "2019-05-21T08:43:02Z",
       "Config": [
         {
           "Path": "/openconfig-system:system/config/hostname",
-          "Value": "localhost:10161"
+          "Value": "localhost-1"
         },
         {
           "Path": "/openconfig-system:system/config/motd-banner",
@@ -212,14 +212,14 @@
         }
       ]
     },
-    "UyPwBSOxDqQM/ZCohHFannSF5ME=": {
-      "ID": "UyPwBSOxDqQM/ZCohHFannSF5ME=",
+    "oWFKpOajlRkQf38YTlCJivn3QlU=": {
+      "ID": "oWFKpOajlRkQf38YTlCJivn3QlU=",
       "Description": "Modification of device sim 2",
       "Created": "2019-05-21T08:43:03Z",
       "Config": [
         {
           "Path": "/openconfig-system:system/config/hostname",
-          "Value": "localhost:10162"
+          "Value": "localhost-2"
         },
         {
           "Path": "/openconfig-system:system/config/motd-banner",
@@ -227,14 +227,14 @@
         }
       ]
     },
-    "5FTJ5QDC0y/RoFnJPfdh8NsB+4Y=": {
-      "ID": "5FTJ5QDC0y/RoFnJPfdh8NsB+4Y=",
+    "1EvL+9i1ZGjKyuSjxh1hlfZ9/74=": {
+      "ID": "1EvL+9i1ZGjKyuSjxh1hlfZ9/74=",
       "Description": "Modification of device sim 3",
       "Created": "2019-05-21T08:43:04Z",
       "Config": [
         {
           "Path": "/openconfig-system:system/config/hostname",
-          "Value": "localhost:10163"
+          "Value": "localhost-3"
         },
         {
           "Path": "/openconfig-system:system/config/motd-banner",

--- a/configs/configStore-sample.json
+++ b/configs/configStore-sample.json
@@ -48,7 +48,7 @@
       "Updated": "2019-05-21T08:43:17Z",
       "Changes": [
         "4cQNf/bxKWdRO0h5GG4a58q5YEQ=",
-        "i86kLjjNFr5IBa+O3YTuv9HBqJg="
+        "UwIDulssBgfXoeKco0eirkKkJUI="
       ]
     },
     "localhost-2-1.0.0": {
@@ -60,7 +60,7 @@
       "Updated": "2019-05-21T08:43:18Z",
       "Changes": [
         "4cQNf/bxKWdRO0h5GG4a58q5YEQ=",
-        "UyPwBSOxDqQM/ZCohHFannSF5ME="
+        "oWFKpOajlRkQf38YTlCJivn3QlU="
       ]
     },
     "localhost-3-1.0.0": {
@@ -72,7 +72,7 @@
       "Updated": "2019-05-21T08:43:19Z",
       "Changes": [
         "4cQNf/bxKWdRO0h5GG4a58q5YEQ=",
-        "5FTJ5QDC0y/RoFnJPfdh8NsB+4Y="
+        "1EvL+9i1ZGjKyuSjxh1hlfZ9/74="
       ]
     },
     "device-1-device-simulator-1.0.0": {

--- a/deployments/helm/onos-config/files/configs/changeStore-sample.json
+++ b/deployments/helm/onos-config/files/configs/changeStore-sample.json
@@ -197,14 +197,14 @@
         }
       ]
     },
-    "i86kLjjNFr5IBa+O3YTuv9HBqJg=": {
-      "ID": "i86kLjjNFr5IBa+O3YTuv9HBqJg=",
+    "UwIDulssBgfXoeKco0eirkKkJUI=": {
+      "ID": "UwIDulssBgfXoeKco0eirkKkJUI=",
       "Description": "Modification of device sim 1",
       "Created": "2019-05-21T08:43:02Z",
       "Config": [
         {
           "Path": "/openconfig-system:system/config/hostname",
-          "Value": "localhost:10161"
+          "Value": "localhost-1"
         },
         {
           "Path": "/openconfig-system:system/config/motd-banner",
@@ -212,14 +212,14 @@
         }
       ]
     },
-    "UyPwBSOxDqQM/ZCohHFannSF5ME=": {
-      "ID": "UyPwBSOxDqQM/ZCohHFannSF5ME=",
+    "oWFKpOajlRkQf38YTlCJivn3QlU=": {
+      "ID": "oWFKpOajlRkQf38YTlCJivn3QlU=",
       "Description": "Modification of device sim 2",
       "Created": "2019-05-21T08:43:03Z",
       "Config": [
         {
           "Path": "/openconfig-system:system/config/hostname",
-          "Value": "localhost:10162"
+          "Value": "localhost-2"
         },
         {
           "Path": "/openconfig-system:system/config/motd-banner",
@@ -227,14 +227,14 @@
         }
       ]
     },
-    "5FTJ5QDC0y/RoFnJPfdh8NsB+4Y=": {
-      "ID": "5FTJ5QDC0y/RoFnJPfdh8NsB+4Y=",
+    "1EvL+9i1ZGjKyuSjxh1hlfZ9/74=": {
+      "ID": "1EvL+9i1ZGjKyuSjxh1hlfZ9/74=",
       "Description": "Modification of device sim 3",
       "Created": "2019-05-21T08:43:04Z",
       "Config": [
         {
           "Path": "/openconfig-system:system/config/hostname",
-          "Value": "localhost:10163"
+          "Value": "localhost-3"
         },
         {
           "Path": "/openconfig-system:system/config/motd-banner",

--- a/deployments/helm/onos-config/files/configs/configStore-sample.json
+++ b/deployments/helm/onos-config/files/configs/configStore-sample.json
@@ -48,7 +48,7 @@
       "Updated": "2019-05-21T08:43:17Z",
       "Changes": [
         "4cQNf/bxKWdRO0h5GG4a58q5YEQ=",
-        "i86kLjjNFr5IBa+O3YTuv9HBqJg="
+        "UwIDulssBgfXoeKco0eirkKkJUI="
       ]
     },
     "localhost-2-1.0.0": {
@@ -60,7 +60,7 @@
       "Updated": "2019-05-21T08:43:18Z",
       "Changes": [
         "4cQNf/bxKWdRO0h5GG4a58q5YEQ=",
-        "UyPwBSOxDqQM/ZCohHFannSF5ME="
+        "oWFKpOajlRkQf38YTlCJivn3QlU="
       ]
     },
     "localhost-3-1.0.0": {
@@ -72,7 +72,7 @@
       "Updated": "2019-05-21T08:43:19Z",
       "Changes": [
         "4cQNf/bxKWdRO0h5GG4a58q5YEQ=",
-        "5FTJ5QDC0y/RoFnJPfdh8NsB+4Y="
+        "1EvL+9i1ZGjKyuSjxh1hlfZ9/74="
       ]
     },
     "device-1-device-simulator-1.0.0": {

--- a/docs/run.md
+++ b/docs/run.md
@@ -25,7 +25,7 @@ go run github.com/onosproject/onos-config/cmd/onos-config \
 ```
 > The plugins here were built locally with a command like
 ```bash
-> CGO_ENABLED=1 go build -o modelplugin/TestDevice-1.0.0/testdevice.so.1.0.0 -buildmode=plugin ./modelplugin/TestDevice-1.0.0
+> CGO_ENABLED=1 go build -o modelplugin/TestDevice-1.0.0/testdevice.so.1.0.0 -buildmode=plugin -tags=modelplugin ./modelplugin/TestDevice-1.0.0
 ```
 > When running with Docker or Kubernetes these plugins will be built and (optionally) loaded
 at startup. To check the list of currently loaded plugins use:
@@ -49,7 +49,7 @@ Here is an example on how to use `gnmi_cli -get` to get configuration for a part
 ```bash
 gnmi_cli -get -address localhost:5150 \
     -proto "path: <target: 'localhost-1', elem: <name: 'openconfig-system:system'> elem:<name:'config'> elem: <name: 'motd-banner'>>" \
-    -timeout 5s -alsologtostderr\
+    -timeout 5s -alsologtostderr \
     -client_crt pkg/southbound/testdata/client1.crt \
     -client_key pkg/southbound/testdata/client1.key \
     -ca_crt pkg/southbound/testdata/onfca.crt

--- a/modelplugin/Devicesim-1.0.0/modelmain.go
+++ b/modelplugin/Devicesim-1.0.0/modelmain.go
@@ -44,7 +44,7 @@ func (m modelplugin) UnmarshalConfigValues(jsonTree []byte) (*ygot.ValidatedGoSt
 	vgs := ygot.ValidatedGoStruct(device)
 
 	if err := devicesim_1_0_0.Unmarshal([]byte(jsonTree), device); err != nil {
-		panic(fmt.Sprintf("Cannot unmarshal JSON: %v", err))
+		return nil, err
 	}
 
 	return &vgs, nil

--- a/modelplugin/ModelGenerator.sh
+++ b/modelplugin/ModelGenerator.sh
@@ -77,7 +77,7 @@ func (m modelplugin) UnmarshalConfigValues(jsonTree []byte) (*ygot.ValidatedGoSt
 	vgs := ygot.ValidatedGoStruct(device)
 
 	if err := $TYPEVERSIONPKG.Unmarshal([]byte(jsonTree), device); err != nil {
-		panic(fmt.Sprintf("Cannot unmarshal JSON: %v", err))
+		return nil, err
 	}
 
 	return &vgs, nil

--- a/modelplugin/TestDevice-2.0.0/modelmain.go
+++ b/modelplugin/TestDevice-2.0.0/modelmain.go
@@ -44,7 +44,7 @@ func (m modelplugin) UnmarshalConfigValues(jsonTree []byte) (*ygot.ValidatedGoSt
 	vgs := ygot.ValidatedGoStruct(device)
 
 	if err := testdevice_2_0_0.Unmarshal([]byte(jsonTree), device); err != nil {
-		panic(fmt.Sprintf("Cannot unmarshal JSON: %v", err))
+		return nil, err
 	}
 
 	return &vgs, nil

--- a/pkg/manager/manager.go
+++ b/pkg/manager/manager.go
@@ -121,6 +121,56 @@ func LoadManager(configStoreFile string, changeStoreFile string, deviceStoreFile
 	return NewManager(&configStore, &changeStore, deviceStore, networkStore, topoChannel)
 }
 
+// ValidateStores validate configurations against their ModelPlugins at startup
+func (m *Manager) ValidateStores() error {
+
+	validationErrors := make(chan error)
+	cfgCount := len(m.ConfigStore.Store)
+	for _, configObj := range m.ConfigStore.Store {
+		go validateConfiguration(configObj, validationErrors)
+	}
+	for valErr := range validationErrors {
+		if valErr != nil {
+			return valErr
+		}
+		cfgCount--
+		if cfgCount == 1 {
+			return nil
+		}
+	}
+	return nil
+}
+
+// validateConfiguration is a go routine for validating a Configuration against it ModelPlugin
+func validateConfiguration(configObj store.Configuration, errChan chan error) {
+	modelPluginName := configObj.Type + "-" + configObj.Version
+	cfgModelPlugin, pluginExists := mgr.ModelRegistry[modelPluginName]
+	if pluginExists {
+		log.Info("Validating config ", configObj.Name, " with Model Plugin ", modelPluginName)
+		fullconfig := configObj.ExtractFullConfig(mgr.ChangeStore.Store, 0)
+		configJSON, err := store.BuildTree(fullconfig)
+		if err != nil {
+			errChan <- err
+			return
+		}
+		ygotModelConfig, err := cfgModelPlugin.UnmarshalConfigValues(configJSON)
+		if err != nil {
+			log.Error(string(configJSON))
+			errChan <- err
+			return
+		}
+		err = cfgModelPlugin.Validate(ygotModelConfig)
+		if err != nil {
+			log.Error(string(configJSON))
+			errChan <- err
+			return
+		}
+	} else {
+		fmt.Println("No Model Plugin available for", modelPluginName)
+	}
+	errChan <- nil
+}
+
 // Run starts a synchronizer based on the devices and the northbound services.
 func (m *Manager) Run() {
 	log.Info("Starting Manager")

--- a/pkg/northbound/diags/diags_test.go
+++ b/pkg/northbound/diags/diags_test.go
@@ -45,7 +45,7 @@ func Test_GetChanges_All(t *testing.T) {
 }
 
 func Test_GetChanges_Change(t *testing.T) {
-	testGetChanges(t, "5FTJ5QDC0y/RoFnJPfdh8NsB+4Y=")
+	testGetChanges(t, "1EvL+9i1ZGjKyuSjxh1hlfZ9/74=")
 }
 
 func testGetChanges(t *testing.T, changeID string) {

--- a/pkg/store/tree.go
+++ b/pkg/store/tree.go
@@ -70,6 +70,9 @@ func addPathToTree(path, value string, nodeif *interface{}) error {
 			(nodemap)[pathelems[1]] = numval
 		} else if strings.ToLower(value) == "true" || strings.ToLower(value) == "false" {
 			(nodemap)[pathelems[1]] = strings.ToLower(value)
+		} else if value == "openconfig-aaa-types:LOCAL" {
+			// FIXME - this is a dirty hack until we create a custom unmarshaller for Paths&Values to YGOT
+			(nodemap)[pathelems[1]] = []string{"openconfig-aaa-types:LOCAL"}
 		} else {
 			(nodemap)[pathelems[1]] = value
 		}
@@ -87,6 +90,7 @@ func addPathToTree(path, value string, nodeif *interface{}) error {
 		listName := pathelems[1][:brktIdx]
 		keyName := pathelems[1][brktIdx+1 : eqIdx]
 		keyVal := pathelems[1][eqIdx+1 : brktIdx2]
+		keyValNum, keyValNumErr := strconv.Atoi(keyVal)
 
 		listSlice, ok := nodemap[listName]
 		if !ok {
@@ -105,7 +109,7 @@ func addPathToTree(path, value string, nodeif *interface{}) error {
 			if !ok {
 				return fmt.Errorf("Failed to convert list slice %d", idx)
 			}
-			if (lsMap)[keyName] == keyVal {
+			if (lsMap)[keyName] == keyVal || (keyValNumErr == nil && (lsMap)[keyName] == keyValNum) {
 				listItemMap = lsMap
 				foundit = true
 				break
@@ -113,7 +117,11 @@ func addPathToTree(path, value string, nodeif *interface{}) error {
 		}
 		if !foundit {
 			listItemMap = make(map[string]interface{})
-			listItemMap[keyName] = keyVal
+			if keyValNumErr == nil {
+				listItemMap[keyName] = keyValNum
+			} else {
+				listItemMap[keyName] = keyVal
+			}
 		}
 		listItemIf := interface{}(listItemMap)
 		err := addPathToTree(refinePath, value, &listItemIf)

